draft-irtf-cfrg-opaque.md

```diff
@@ -393,152 +391,346 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
 We let `Npk` and `Nsk` denote the size of public and private keys, respectively,
 used in the AKE.
 
-# Offline Registration {#offline-phase}
+- Deterministic AKE Key Generation Functions:
+  - RecoverPublicKey(private_key): Recover the public key related to the input private key.
+  - DeriveAkeKeyPair(seed): Derive a private and public key pair deterministically from the seed.
 
-Registration is executed between a client C and a
-server S. It is assumed S can identify C and the client can
-authenticate S during this registration phase. This is the only part
-in OPAQUE that requires an authenticated and confidential channel, either physical, out-of-band,
-PKI-based, etc. This section describes the registration flow, message encoding,
-and helper functions. Moreover, C has a key pair (client_private_key, client_public_key) for an AKE protocol
-which is suitable for use with OPAQUE; See {{online-phase}}. The private-public keys (client_private_key, client_public_key) may be randomly generated (using a cryptographically secure pseudorandom number generator) for the account or provided by the calling client.
-Clients MUST NOT use the same key pair (client_private_key, client_public_key) for two different accounts.
+Random nonces used in this protocol are of length `Nn = 32` bytes.
 
-## Setup Phase {#setup-phase}
+We define `nil` as a lack of value.
 
-In a setup phase, C chooses its password, and S chooses its own pair of private-public
-AKE keys (server_private_key, server_public_key) for use with the AKE, along with a Nh-byte oprf_seed. S can use
-the same pair of keys with multiple clients, and can opt to use multiple seeds (so long as they are
-kept consistent for each client). These steps can happen offline, i.e., before the registration phase.
+# Client Credential Storage {#client-credential-storage}
 
-Once complete, the registration process proceeds as follows.
+OPAQUE makes use of a structure `Envelope` to manage client credentials. This envelope holds information about its format and content for the client to obtain its authentication material.
 
-## Credential Registration
+OPAQUE allows applications to either provide custom client private and public keys for authentication or to generate them internally, making the application oblivious to the client's private key. Each public and private key value is an opaque byte string, specific to the AKE protocol in which OPAQUE is instantiated.
+
+These two options are defined as the `Internal Key` and `External Key` modes.  See {{envelope-modes}} for their specifications.
```

Comment:
Can we call this "Internal" and "External", rather than "Internal Key" and "External Key"?